File: firmware/src/recurrence.cpp
#include "recurrence.h"
#include <time.h>

time_t nextOccurrence(time_t original_scheduled_at, Recurrence r) {
    time_t now = time(nullptr);
    time_t next = original_scheduled_at;

    switch (r) {
        case Recurrence::Daily:
            while (next <= now) next += 86400;
            break;

        case Recurrence::Weekly:
            while (next <= now) next += 7 * 86400;
            break;

        case Recurrence::Weekdays: {
            while (next <= now) {
                next += 86400;
                struct tm tm;
                localtime_r(&next, &tm);
                // Skip Saturday (6) and Sunday (0)
                while (tm.tm_wday == 0 || tm.tm_wday == 6) {
                    next += 86400;
                    localtime_r(&next, &tm);
                }
            }
            break;
        }

        default:
            break;
    }
    return next;
}

File: web/src/lib/server/recurrence.ts
import type { Recurrence } from '$lib/types';

export function nextOccurrence(scheduledAt: string, recurrence: Recurrence): string {
  const now = new Date();
  let next = new Date(scheduledAt);

  switch (recurrence) {
    case 'daily':
      while (next <= now) {
        next.setDate(next.getDate() + 1);
      }
      break;

    case 'weekly':
      while (next <= now) {
        next.setDate(next.getDate() + 7);
      }
      break;

    case 'weekdays':
      while (next <= now) {
        next.setDate(next.getDate() + 1);
        // Skip Saturday (6) and Sunday (0)
        while (next.getDay() === 0 || next.getDay() === 6) {
          next.setDate(next.getDate() + 1);
        }
      }
      break;

    default:
      break;
  }

  return next.toISOString();
}
